src/Sentry/Internal/MemoryMonitor.cs

```diff
@@ -0,0 +1,164 @@
+
+        _options.HeapDumpDebouncer.RecordOccurence(eventTime);
+
+        var usedMemoryPercentage = ((double)usedMemory / _totalMemory) * 100;
+        _options.LogDebug("Total Memory: {0:N0} bytes", _totalMemory);
+        _options.LogDebug("Memory used: {0:N0} bytes ({1:N2}%)", usedMemory, usedMemoryPercentage);
+        _options.LogDebug("Automatic heap dump triggered");
+        _onCaptureDump();
+    }
+
+    internal void CaptureMemoryDump()
+    {
+        if (_options.DisableFileWrite)
+        {
+            _options.LogDebug("File write has been disabled via the options. Unable to create memory dump.");
+            return;
+        }
+
+        var dumpFile = TryGetDumpLocation();
+        if (dumpFile is null)
+        {
+            return;
+        }
+
+        var processId = Environment.ProcessId;
+        _options.LogInfo("Creating a memory dump for Process ID: {0}", processId);
+
+        // Check which patth to use for dotnet-gcdump. If it's been bundled with the application, it will be available
+        // in the `dotnet-gcdump` folder of the application directory. Otherwise we assume it has been installed globally.
+        var bundledToolPath = Path.Combine(AppContext.BaseDirectory, "dotnet-gcdump", "dotnet-gcdump.dll");
+        if (File.Exists(bundledToolPath))
+        {
+            _options.LogDebug($"Using bundled version of dotnet-gcdump from: {bundledToolPath}");
+        }
+        else
+        {
+            _options.LogDebug("Using global version of dotnet-gcdump");
+        }
+
+        var arguments = $"collect -p {processId} -o '{dumpFile}'";
+        var command = File.Exists(bundledToolPath)
+            ? $"dotnet {bundledToolPath} {arguments}"
+            : $"dotnet-gcdump {arguments}";
+        var startInfo = new ProcessStartInfo
+        {
+            FileName = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "cmd.exe" : "/bin/bash",
+            Arguments = $"-c \"{command}\"",
+            RedirectStandardOutput = true,
+            UseShellExecute = false,
+            CreateNoWindow = true,
+        };
+
+        var process = new Process { StartInfo = startInfo };
+        process.Start();
+        while (!process.StandardOutput.EndOfStream)
+        {
+            if (process.StandardOutput.ReadLine() is { } line)
+            {
+                _options.LogDebug(line);
+            }
+        }
+
+        if (!_options.FileSystem.FileExists(dumpFile))
+        {
+            // if this happens, hopefully there would be more information in the standard output from the process above
+            _options.LogError("Unexpected error creating memory dump. Check debug logs for more information.");
+        }
+
+        _onDumpCollected(dumpFile);
+    }
+
+    internal string? TryGetDumpLocation()
+    {
+        try
+        {
+            var rootPath = _options.CacheDirectoryPath ??
+                           Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
```

Comment:
I believe (could be wrong) this can actually create a directory.

See: https://github.com/dotnet/runtime/blob/dfc2b85a00f3baf5ac52d7615bf857e6217011c8/src/libraries/System.Private.CoreLib/src/System/Environment.GetFolderPathCore.Unix.cs#L44

So we'd need to guard this under the `IFileSystem` abstraction, and/or check `options.DisableFileWrite` option before calling this.

Reply:
Damned side affects - the F# community would be ashamed. 

In any event, I've added this in the Hub (when instantiating the MemoryMonitor):
https://github.com/getsentry/sentry-dotnet/blob/2e6a0f131317fff04658f93aa0815d585b31f7d6/src/Sentry/Internal/Hub.cs#L71-L78